### ISSUE TITLE
[Feature request

### ISSUE THREAD

**Issue by nebeker** [chummer5a/chummer5a#4425]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Comment by nebeker** [comment]
Here's wha I'm working on: [link to branch](https://github.com/nebeker/chummer5a/tree/larger-text-sizes).

**Comment by JoschiGrey** [comment]
Why not making your larger fonts setting a variable instead of just an checkbox. That way the user has more control. 
A numeric up down could work well. That way you can set a min and max value to prevent someone to input some stupid stuff.

**Comment by nebeker** [comment]
That would be the ideal UX, I think, but you'd need either a single "change text size by X %" option or separate options for text sizes for different control categories. Unless everything is the same size - is everything the same size? That would allow for a single "UI text size; X pt" option.

Otherwise, as far as mim/max limits, I can't imagine anybody wanting smaller type and this 50% increase feels like the sweet spot between legibility and not breaking the UI too much.

**Comment by JoschiGrey** [comment]
Yeah I was thinking about an increase by X% numupdown with the without allowing negative values.
Set the default to 50% as a suggestion and keep the on/off checkbox.

I think most if not all controls use the same text size, but your solution with relative size increases seems better anyway that inputting your own sizes.

Making the whole UI more adaptive to bigger text sizes would be great for making Chummer more accessible to vision impaired users, but that is a ton of work.

**Comment by nebeker** [comment]
> Making the whole UI more adaptive to bigger text sizes would be great for making Chummer more accessible to vision impaired users, but that is a ton of work.

Exactly. I was thinking of the 20/80 rule here. This makes it much more useable, without as little effort as possible.

How do you feel about the performance impact of going through all the controls when loading a new form?

**Comment by DelnarErsike** [comment]
So, bad news bears, this type of thing *should* be handled automatically by the weeks of work I put into trying to wrangle high DPI scaling, but it inevitably is not possible while Chummer is based on WinForms and the Multiple Document Interface (MDI) system. While you can try to do font manipulation as a stopgap measure, there are many, many other places where doing only font manipulations will cause things to break or simply not be good enough. We really need proper DPI scaling handling, and that just flatout is not possible due to a known bug in the systems on which Chummer relies.

If you want a proper solution, you'd need to migrate Chummer to a completely new system. Granted, it's a long time coming anyway because aging WinForms is also the number 1 reason why loading characters is so slow, but it's a massive amount of work that the current devs simply don't have the patience to tackle.

**Comment by JoschiGrey** [comment]
Migrating Chummer to a completely different system? Damn that sounds painful at best.
Just out of curiosity which system would you guys transfer to if you would takle that problem?

**Comment by DelnarErsike** [comment]
WPF would be the least painful, and even that has been quite painful. The `WPF-UI-Rewrite` branch was last updated in 2018. Something like Xamarin would be more painful, but allow for cross-platform stuff.

**Comment by JoschiGrey** [comment]
What does that all include?

- Recreating all the forms in WPF
- Re linking all the forms and controls
- Refactoring some code that is `WinForm` specific over to `WPF`

**Comment by nebeker** [comment]
Could we still do it this way and allow users to toggle an "experimental" option? I'd be OK with messing up a few things to be able to read text more easily. Or is the risk of things breaking unexpectedly for inattentive users too great?

**Comment by ArchonMegalon** [comment]
VS 2022 will include the new MAUI Framework, which kind of is an evolved
version of Xamarin.Forms.

Would run on Android and Macs as well...

https://visualstudiomagazine.com/articles/2021/01/05/maui.aspx

**Comment by nebeker** [comment]
Yeah, MAUI can't come soon enough.

This is basically going to be incidentally added with a port, then, right?

**Comment by DelnarErsike** [comment]
I mean, it will eventually be added when we finally switch out from WinForms.
You can keep working on your PR if you want, it's just that you'll probably end up running into similar issues with spacing scaling if you only take care of fonts... and if you want to take care of both fonts and spacing, you'd kind of be rewriting the entire WinForms DPI scaling system.

Me personally, I'd just avoid doing duplicate work and handle this properly through using a platform with working DPI scaling. In the meantime, I'll continue to advise people to switch out their High DPI Scaling settings for Chummer to System or System (Enhanced), maybe even set System as the new default instead of PerMonitorV2 (System Enhanced has a problems with a lot of machines and so I would not set it as the default).

**Comment by nebeker** [comment]
Fair enough. No point reinventing the wheel if the new 19 inch chrome plated rims are on the way.

I did notice size 4pt text when using a different monitor. It really is a DPI issue. Thanks for suggesting workarounds.

**Comment by JoschiGrey** [comment]
> I mean, it will eventually be added when we finally switch out from WinForms.
So waiting for MAUI and then "we bite into the sauer apple" as we say in German ;)

But implementing this blanket text size option with an warning that it could (and probably will) break some UI elements still seems like an good option. The normal user won't be affected and it may help some people.

**Comment by DelnarErsike** [comment]
In the meantime, have a look at Nightlies compiled after commit fff0b6750fe5e8c2f4c9320ef61727a2694a54a0. 5.214.63 should be the first one. I tweaked some stuff regarding DPI scaling and exposed the selectable method to the Global Options screen. From my limited testing, it seems to work quite well.

**Comment by nebeker** [comment]
> In the meantime, have a look at Nightlies compiled after commit [fff0b67](https://github.com/chummer5a/chummer5a/commit/fff0b6750fe5e8c2f4c9320ef61727a2694a54a0). 5.214.63 should be the first one. I tweaked some stuff regarding DPI scaling and exposed the selectable method to the Global Options screen. From my limited testing, it seems to work quite well.

That sounds great. I'll check it out as soon as it's available.

**Comment by nebeker** [comment]
I've played around with the new settings on a 1080p monitor at 100% scaling and didn't notice any difference in the text size. I imagine that's because it doesn't need any scaling in this scenario.

So... This seems like partially a DPI issue, but also an issue with the baseline text size. Thoughts?

**Comment by DelnarErsike** [comment]
> 
> 
> I've played around with the new settings on a 1080p monitor at 100% scaling and didn't notice any difference in the text size. I imagine that's because it doesn't need any scaling in this scenario.
> 
> So... This seems like partially a DPI issue, but also an issue with the baseline text size. Thoughts?

The default font size used by Chummer is 8.25 pt Microsoft Sans Serif, which is the exact same size (or very similar size) that you'll find in a lot of other Windows programs. You can even see this in your screenshots in the original issue message, where if you compare the default Chummer font size to the font size of text in Chummer's window title, they are almost identical. If you can handle every program except Chummer at 100%, that's a bit odd, but you can change your DPI settings on a per-program basis in Windows 10 if you need to.

The DPI issue is usually what people talk about when they say that Chummer's text is too small on their screen because DPI scaling is what people use to make text and items appear larger without decreasing their screen resolution. The original approach I had set up for Chummer's DPI scaling was PerMonitorV2 because that's supposed to be the one that handles things the most thoroughly and with the least amount of blur. However, as it turns out, PerMonitorV2 has issues when running in combination with the Multiple Document Interface system, which is what Chummer uses for tabbing around between characters, the character roster, and the master index. The commit I mentioned did two things: switch the default method to System, which just zooms everything in, and manually add in DPI scaling mode code into Chummer instead of relying on settings in the manifest file. The latter also let us expose DPI scaling modes in Chummer's Options screen, making it more likely that people would pick up on it compared to looking in the Compatibility tab of the Chummer exe.
